<a href="https://colab.research.google.com/github/FelipeGonzalezGarrido/Coding-Dojo/blob/main/Simulacion_de_examen_pendiente_grafico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [58]:
from google.colab import drive
drive.mount ("/content/drive")
import pandas as pd

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [59]:
filename = "/content/drive/MyDrive/Data CodingDojo/literacy_rates.csv"

df = pd.read_csv(filename)
df.head()

Region      Country  Year    Age  Gender Literacy rate
0  Central and Southern Asia  Afghanistan  2011    15+  female     0.1761206
1  Central and Southern Asia  Afghanistan  2011    15+    male      0.454171
2  Central and Southern Asia  Afghanistan  2011    15+   total     0.3174112
3                        NaN  Afghanistan  2011  15-24  female     0.3211322
4                        NaN  Afghanistan  2011  15-24    male     0.6187907

# Info dataframe
#### 4955 registros y 6 columnas
#### presenta 5 nulos en la columna Region
#### Sin Registros duplicados

In [60]:
df.shape

(4955, 6)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4955 entries, 0 to 4954
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Region         4950 non-null   object
 1   Country        4955 non-null   object
 2   Year           4955 non-null   int64 
 3   Age            4955 non-null   object
 4   Gender         4955 non-null   object
 5   Literacy rate  4955 non-null   object
dtypes: int64(1), object(5)
memory usage: 232.4+ KB


In [62]:
df.isna().sum()

Region           5
Country          0
Year             0
Age              0
Gender           0
Literacy rate    0
dtype: int64

In [63]:
df[df["Region"].isna()]

Region      Country  Year    Age  Gender       Literacy rate
3      NaN  Afghanistan  2011  15-24  female           0.3211322
4      NaN  Afghanistan  2011  15-24    male           0.6187907
464    NaN   Uzbekistan  2016    65+  female  0.9966974999999999
465    NaN   Uzbekistan  2016    65+    male           0.9990176
466    NaN   Uzbekistan  2016    65+   total  0.9977187999999999

Se detecta que registros con columna region en null pertenecen a Central and Southern Asia. por ende se llenaran los null con aquel valor

In [64]:
df[df["Country"] == "Afghanistan"].head()

Region      Country  Year    Age  Gender Literacy rate
0  Central and Southern Asia  Afghanistan  2011    15+  female     0.1761206
1  Central and Southern Asia  Afghanistan  2011    15+    male      0.454171
2  Central and Southern Asia  Afghanistan  2011    15+   total     0.3174112
3                        NaN  Afghanistan  2011  15-24  female     0.3211322
4                        NaN  Afghanistan  2011  15-24    male     0.6187907

#1. Sin registros duplicados en dataframe

In [65]:
df.duplicated().sum()

0

#2. Identificar y abordar cualquier dato que falte en este conjunto de datos. Trátenlos de forma adecuada. Incluyan una breve descripción en el método que suelen tratar con los valores que faltan junto con una justificación para el uso de ese método.

In [66]:
# En este caso se busco de forma individual, por pais, para encontrar el dato faltante.
# como ambos registros nulos pertenecen a la misma region se opto por llenar el dato con la region que corresponde.
df["Region"].fillna("Central and Southern Asia", inplace = True)

#3. Asegúrense de que todas las columnas coincidan con los tipos de datos enumerados en el diccionario de datos.
La columna Literacy rate tiene formato object, debe ser float.

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4955 entries, 0 to 4954
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Region         4955 non-null   object
 1   Country        4955 non-null   object
 2   Year           4955 non-null   int64 
 3   Age            4955 non-null   object
 4   Gender         4955 non-null   object
 5   Literacy rate  4955 non-null   object
dtypes: int64(1), object(5)
memory usage: 232.4+ KB


In [76]:
df["Literacy rate"] = df["Literacy rate"].astype(float)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4955 entries, 0 to 4954
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Region         4955 non-null   object 
 1   Country        4955 non-null   object 
 2   Year           4955 non-null   int64  
 3   Age            4955 non-null   object 
 4   Gender         4955 non-null   object 
 5   Literacy rate  4955 non-null   float64
dtypes: float64(1), int64(1), object(4)
memory usage: 232.4+ KB


# En primera instancia no se permite la conversion de data type de la columna Literacy rate ya que tiene un valor en porcentaje en lugar de decimal

In [69]:
df[(df["Country"] == "Afghanistan")& (df["Age"] == "25-64")]

Region      Country  Year    Age  Gender  \
6   Central and Southern Asia  Afghanistan  2011  25-64  female   
7   Central and Southern Asia  Afghanistan  2011  25-64    male   
8   Central and Southern Asia  Afghanistan  2011  25-64   total   
17  Central and Southern Asia  Afghanistan  2018  25-64  female   
18  Central and Southern Asia  Afghanistan  2018  25-64    male   
19  Central and Southern Asia  Afghanistan  2018  25-64   total   

          Literacy rate  
6   0.08412760000000001  
7             0.3779481  
8             0.2315362  
17            0.1433139  
18              45.384%  
19  0.30457850000000003

In [75]:
df[(df["Country"] == "Afghanistan")& (df["Age"] == "25-64")]

Region      Country  Year    Age  Gender  \
6   Central and Southern Asia  Afghanistan  2011  25-64  female   
7   Central and Southern Asia  Afghanistan  2011  25-64    male   
8   Central and Southern Asia  Afghanistan  2011  25-64   total   
17  Central and Southern Asia  Afghanistan  2018  25-64  female   
18  Central and Southern Asia  Afghanistan  2018  25-64    male   
19  Central and Southern Asia  Afghanistan  2018  25-64   total   

          Literacy rate  
6   0.08412760000000001  
7             0.3779481  
8             0.2315362  
17            0.1433139  
18              0.45384  
19  0.30457850000000003

In [74]:
#Cambio de valor 45,384% a 0,45384 
#df.loc[18,"Literacy rate"] = "0.45384"

#4. Identifiquen y aborden cualquier inconsistencias en los valores categóricos (ejemplo: cat, Cat, cats).

Se valida con funcion . value_counts() en columnas sin encontrar datos categoricos incosistentes.

In [81]:
df.Region.value_counts()

Latin America and the Caribbean     1410
Sub-Saharan Africa                  1155
Northern Africa and Western Asia     852
Europe and Northern America          504
Central and Southern Asia            467
Eastern and South-Eastern Asia       423
Oceania                              144
Name: Region, dtype: int64

In [82]:
df.Gender.value_counts()

female    1652
total     1652
male      1651
Name: Gender, dtype: int64

#5. Identifiquen y aborden cualquier punto de datos inapropiados o inusuales (ejemplo: la edad no puede ser 325).
Se encontro este problema en Literacy rate de afganistan, correspondiente al porcentaje de hombres de 25 a 64 a;os.
el valor es 45.384 % mientras que los demas valores estan expresados en decimal.
se realiza el cambio de 45.384 % a 0.45384 que se ajusta a lo presentado en a;os anteriores.

In [83]:
df[(df["Country"] == "Afghanistan")& (df["Age"] == "25-64")]

Region      Country  Year    Age  Gender  Literacy rate
6   Central and Southern Asia  Afghanistan  2011  25-64  female       0.084128
7   Central and Southern Asia  Afghanistan  2011  25-64    male       0.377948
8   Central and Southern Asia  Afghanistan  2011  25-64   total       0.231536
17  Central and Southern Asia  Afghanistan  2018  25-64  female       0.143314
18  Central and Southern Asia  Afghanistan  2018  25-64    male       0.453840
19  Central and Southern Asia  Afghanistan  2018  25-64   total       0.304579

#2 GRAFICOS PENDIENTES